digest/src/lib.rs

```diff
@@ -0,0 +1,39 @@
+pub use multihash;
+use multihash::Multihash;
+
+const SEPARATOR: &[u8; 1] = b":";
+
+pub fn digest(bytes: &[u8]) -> Multihash {
+    multihash::Sha3_256::digest(bytes)
+}
+
+pub trait RootDigest {
+    fn digest_root(self) -> Multihash;
+}
+
+pub trait FieldDigest {
+    fn digest_field(self, field_name: String) -> Multihash;
+}
+
+impl FieldDigest for String {
+    fn digest_field(self, field_name: String) -> Multihash {
+        let mut bytes = field_name.into_bytes();
```

Comment:
I get your point regarding encoding however I don't see how this helps with the hash of the actual value.

Reply:
If we have full control over what is hashed by providing the exact bytes that act as the identifier for this field, we are less prone to bugs due to renaming etc.

It will also allows us for better interoperability at some point where someone in a different language has to implement this algorithm and be complaint with our implementation.

Regarding the actual hash of the value:

Maybe we should **not** be implementing digest on any primitive values. I think it is reasonable to expect consumers of `Digest` to implement the trait themselves and not use primitive types.

Reply:
Thinking more about it, I still do not understand why you are pushing for a hardcoded byte value for fields.

The problematic that you are raising is the same for field names and `String` values as we are processing field names as `String`.

If the lack of encoding specification is an issue then this issue affect both sides. Because of that, I don't see how a hardcoded byte value per field would help as it would not solve the issue on the value side.

`String::as_bytes` returns the bytes for UTF-8. I think it is fair to just specify that the digest of a field is the hash of `<name of the field as bytes, utf-8 encoded>:<value as bytes, utf-8 encoded>`.

Reply:
> Maybe we should **not** be implementing digest on any primitive values. I think it is reasonable to expect consumers of `Digest` to implement the trait themselves and not use primitive types.

I see.

Reply:
How do you expect the `FieldDigest` implementation of bitcoin mainnet network to look like? I was going with the utf-8 bytes of `mainnet` string but you seem to be thinking something else.

Reply:
Well, the bitcoin mainnet network actually has an unambigous byte representation: the 4 magic bytes in front of every message. That is something I would have used.

In the context of Ethereum, probably the bytes of the chain id in big/little endian, whatever we specify.

Reply:
> Maybe we should **not** be implementing digest on any primitive values. I think it is reasonable to expect consumers of `Digest` to implement the trait themselves and not use primitive types.

We are adding a level of complexity here where the consumer would need to implement a foreign trait on possible foreign types.

Reply:
I would say the opposite: Computing the digest correctly is such a crucial thing that you should be forced to implement this on your own structs and not use foreign ones.

Reply:
Proposal implemented.